repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletResponseWrapper.scala

```diff
@@ -63,67 +76,275 @@ class HttpServletResponseWrapper(originalResponse: HttpServletResponse, headerMo
-   */
-  override def appendHeader(headerName: String, headerValue: String): Unit = ???
+  private val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+  private lazy val bodyPrintWriter = new PrintWriter(bodyOutputStream)
+
+  private var headerMap: Map[String, Seq[String]] = new TreeMap[String, Seq[String]]()(caseInsensitiveOrdering)
+  private var responseBodyType = ResponseBodyType.Available
+
+  override def getResponse: ServletResponse = throw new UnsupportedOperationException("getResponse is not supported")
+
+  override def setResponse(servletResponse: ServletResponse): Unit =
+    throw new UnsupportedOperationException("setResponse is not supported")
+
+  override def getHeaderNamesList: util.List[String] = getHeaderNames.toList
+
+  override def getHeaderNames: util.Collection[String] = headerMap.keys
+
+  override def containsHeader(name: String): Boolean = headerMap.contains(name)
+
+  override def getHeader(name: String): String = getHeaderValues(name).headOption.orNull
+
+  override def getHeaders(name: String): util.Collection[String] = getHeaderValues(name)
+
+  override def getHeadersList(name: String): util.List[String] = getHeaderValues(name)
+
+  override def getPreferredHeaders(name: String): util.List[String] = getPreferredHeaderValues(getHeaderValues(name))
+
+  override def getPreferredSplittableHeaders(name: String): util.List[String] =
+    getPreferredHeaderValues(getSplittableHeaders(name))
+
+  override def getSplittableHeaders(name: String): util.List[String] =
+    getHeaderValues(name).foldLeft(List.empty[String])((list, value) => list ++ value.split(","))
+
+  override def addHeader(name: String, value: String): Unit = {
+    headerMap = headerMap + (name -> (headerMap.getOrElse(name, Seq.empty[String]) :+ value))
+
+    if (headerMode != ResponseMode.MUTABLE) {
+      // Write through to the wrapped response immediately
+      super.addHeader(name, value)
+    }
+  }
+
+  override def addHeader(name: String, value: String, quality: Double): Unit =
+    addHeader(name, s"$value;q=$quality")
+
+  override def addIntHeader(name: String, value: Int): Unit = addHeader(name, value.toString)
 
   /**
-   * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
+   * Formats the input time (as milliseconds since the epoch) to a format defined in RFC2616.
    */
-  override def appendHeader(headerName: String, headerValue: String, quality: Double): Unit = ???
+  override def addDateHeader(name: String, timeSinceEpoch: Long): Unit =
+    addHeader(name, DateUtils.formatDate(new Date(timeSinceEpoch)))
 
   /**
    * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
    */
-  override def removeHeader(headerName: String): Unit = ???
+  override def appendHeader(name: String, value: String): Unit = {
+    ifMutable(headerMode) {
+      val existingValues = getHeaders(name)
+      existingValues.lastOption match {
```

Comment:
I've made the change and will push it soon.

I'd still like to know the reason again, if you happen to remember.